File: leetcode/1861-rotating-the-box/solution.py
from typing import List


class Solution:
    def rotateTheBox(self, box: List[List[str]]) -> List[List[str]]:
        m, n = len(box), len(box[0])
        grid = [row[:] for row in box]

        # gravity pulls stones ('#') to the right in each row; obstacles
        # ('*') block them and act as new anchor points.
        for row in grid:
            write = n - 1
            for col in range(n - 1, -1, -1):
                if row[col] == '*':
                    write = col - 1
                elif row[col] == '#':
                    row[col] = '.'
                    row[write] = '#'
                    write -= 1

        # rotate 90 degrees clockwise: new[i][j] = old[m-1-j][i]
        rotated = [['.' for _ in range(m)] for _ in range(n)]
        for i in range(m):
            for j in range(n):
                rotated[j][m - 1 - i] = grid[i][j]
        return rotated

File: leetcode/1861-rotating-the-box/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<char>> rotateTheBox(vector<vector<char>>& box) {
        int m = box.size(), n = box[0].size();
        vector<vector<char>> grid = box;

        for (auto& row : grid) {
            int write = n - 1;
            for (int col = n - 1; col >= 0; col--) {
                if (row[col] == '*') {
                    write = col - 1;
                } else if (row[col] == '#') {
                    row[col] = '.';
                    row[write] = '#';
                    write--;
                }
            }
        }

        vector<vector<char>> rotated(n, vector<char>(m, '.'));
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                rotated[j][m - 1 - i] = grid[i][j];
            }
        }
        return rotated;
    }
};
